src/wayland/compositor/tree.rs

```diff
@@ -74,11 +74,12 @@ where
         }
         // orphan all our children
         for child in &my_data.children {
-            // don't do anything if this child is ourselves
-            if child.as_ref().equals(surface.as_ref()) {
+            let child_mutex = child.as_ref().user_data::<Mutex<SurfaceData<R>>>().unwrap();
+            if child_mutex as *const _ == my_data_mutex as *const _ {
```

Comment:
```suggestion
            if std::ptr::eq(child_mutex, my_data_mutex) {
```

I think this would be better style, while keeping the intent clear.